lib/bson/document.rb

```diff
@@ -269,25 +269,24 @@ def dig(*keys)
       end
     end
 
-    if instance_methods.include?(:slice)
-      # Slices a document to include only the given keys.
-      # Will normalize symbol keys into strings.
-      # (this method is backported from ActiveSupport::Hash)
-      #
-      # @example Get a document/hash with only the `name` and `age` fields present
-      #   document # => { _id: <ObjectId>, :name => 'John', :age => 30, :location => 'Earth' }
-      #   document.slice(:name, 'age') # => { name: 'John', age: 30 }
-      #   document.slice('name') # => { name: 'John' }
-      #   document.slice(:foo) # => nil
-      #
-      # @param [ Array<String, Symbol> ] *keys Keys, that will be kept in the resulting document
-      #
-      # @return [ BSON::Document ] The document with only the selected keys
-      #
-      # @since 4.3.1
-      def slice(*keys)
-        super(*keys.map{|key| convert_key(key)})
-      end
+    # Slices a document to include only the given keys.
+    # Will normalize symbol keys into strings.
+    # (this method is backported from ActiveSupport::Hash)
+    #
+    # @example Get a document/hash with only the `name` and `age` fields present
+    #   document # => { _id: <ObjectId>, :name => "John", :age => 30, :location => "Earth" }
+    #   document.slice(:name, 'age') # => { "name": "John", "age" => 30 }
+    #   document.slice('name') # => { "name" => "John" }
+    #   document.slice(:foo) # => {}
+    #
+    # @param [ Array<String, Symbol> ] *keys Keys, that will be kept in the resulting document
+    #
+    # @return [ BSON::Document ] The document with only the selected keys
+    #
+    # @since 4.3.1
+    def slice(*keys)
+      keys.map{|key| convert_key(key)}
+          .each_with_object(BSON::Document.new) {|key, hash| hash[key] = self[key] if has_key?(key)}
```

Comment:
What is the purpose of has_key?

Reply:
It ensures we only return a subset of the keys of the document and dont add new keys by adding additional params with nil values to the document through the slice method. I'll add a test case for this